translationUtils.ts

```diff
@@ -0,0 +1,284 @@
+			});
+		};
+
+		// Handle various things that can go wrong when loading an image.
+		// This includes network issues, dead links, etc.
+		function onFailed() {
+			resolve(null);
+		}
+		image.onerror = onFailed;
+
+		image.src = imageUrl.toString();
+	});
+}
+
+function checkIsSupportedLanguage(translation: TranslationResult) {
+	const supportedLanguages = ['jp', 'kr', 'cn'];
+	return supportedLanguages.includes(translation.originalLanguage);
+}
+
+function overlayTranslations(
+	imageBase64: string,
+	translations: TranslationResult[]
+): Promise<ImageBase64> {
+	return new Promise(resolve => {
+		const image = new Image();
+		image.onload = function () {
+			const canvas = createCanvas(image.width, image.height);
+			const context = canvas.getContext('2d');
+			context.drawImage(
+				image,
+				0,
+				0,
+				image.width,
+				image.height,
+				0,
+				0,
+				image.width,
+				image.height
+			);
+
+			const filteredTranslations = translations.filter(x => x.translatedText.trim() !== '');
+
+			// Draw text bubble as a rounded rectangle.
+			for (const translation of filteredTranslations) {
+				const width = translation.maxX - translation.minX;
+				const height = translation.maxY - translation.minY;
+				const radius = 10; // Roughly 10px.
+				const backgroundColor = 'white';
+
+				drawRectangle(
+					context,
+					translation.minX,
+					translation.minY,
+					width,
+					height,
+					radius,
+					backgroundColor
+				);
+			}
+
+			// Write text on the bubbles.
+			for (const translation of filteredTranslations) {
+				const maxLineWidth = translation.maxX - translation.minX;
+				const maxHeight = translation.maxY - translation.minY;
+				const startY = translation.minY;
+				const maxFontSizePx = 300; // Arbitrarily large starting font.
+
+				fitText(
+					context,
+					translation.translatedText,
+					translation.minX,
+					startY,
+					maxHeight,
+					maxLineWidth,
+					maxFontSizePx
+				);
+			}
+
+			resolve(canvas.toDataURL('image/png'));
+		};
+
+		image.src = imageBase64;
+	});
+}
+
+// Modified from Grumdrig's StackOverflow response.
+function drawRectangle(
+	context: CanvasRenderingContext2D,
+	x: number,
+	y: number,
+	width: number,
+	height: number,
+	radius: number,
+	backgroundColor?: string
+) {
+	if (width < 2 * radius) {
+		radius = width / 2;
+	}
+	if (height < 2 * radius) {
+		radius = height / 2;
+	}
+
+	context.beginPath();
+	context.moveTo(x + radius, y);
+	context.arcTo(x + width, y, x + width, y + height, radius);
+	context.arcTo(x + width, y + height, x, y + height, radius);
+	context.arcTo(x, y + height, x, y, radius);
+	context.arcTo(x, y, x + width, y, radius);
+	context.closePath();
+
+	if (backgroundColor) {
+		context.fillStyle = backgroundColor;
+		context.fill();
+	}
+}
+
+function fitText(
+	context: CanvasRenderingContext2D,
+	text: string,
+	offsetX: number,
+	startY: number,
+	maxHeight: number,
+	maxLineWidth: number,
+	maxFontSizePx: number
+) {
+	if (text === '') {
```

Comment:
Note I'm handling the empty string edge case instead of throwing. Why? Because the contract of the function looks like it can handle any arbitrary string. The empty string is a completely valid string, so it seems reasonable we should handle it instead of throwing.